working/static metaprogramming/intro.md

```diff
@@ -0,0 +1,431 @@
+  more targeted code generators for each of the desired features (`copyWith()`,
+  `==`, `hashCode`, etc).
+* **Generated code must be debuggable and user visible**. See the Usability
+  section below.
+
+## Design constraints
+
+In addition to user-level requirements, there are some practical constraints to
+ensure that we don't compromise the existing advantages of Dart.
+
+* **Must support modular compilation**. Any individual library should be able to
+  be fully compiled (including running all macros) using only the transitive
+  dependencies of the library. To this end we should not expose any API or
+  functionality that allows global access to the program, outside of the current
+  library and/or its dependencies. Such functionality would defeat modular
+  compilation.
+* **Must support incremental compilation**. This is key to the hot reload
+  workflow—any change in the program should not have to rerun all
+  metaprogramming in the entire program. It should at worst require a rerun of
+  the metaprogramming in libraries that depend on the modified library.
+* **Must support expression evaluation**. Expression evaluation is a vital tool
+  for the development experience and we need to ensure it is not negatively
+  affected.
+
+For the most part the design constraints here align with the modular compilation
+design constraints. In short, metaprogramming should not introduce global action
+at a distance.
+
+## Non-goals
+
+* **Modifying existing code**. We don’t see a need to allow modifiying existing
+  code through metaprogramming. This leads to unexpected behavior, and
+  negatively affects usability. Dart code that already has clearly defined
+  meaning today should continue to mean the same thing after metaprogramming.
+  The goal is not to be able to hijack existing Dart syntax to do wildly
+  different things, but to augment code with new capabilities.
+
+## Approach
+
+To understand what a static metaprogramming solution might look like, we’ll walk
+through one of the use cases. Say a Dart user wants to write a library that
+automatically defines `copyWith()` methods for classes. For them to do that,
+Dart needs to provide a way for users to:
+
+1. **Invoke the metaprogramming** to control which classes should get the
+   `copyWith()` method.
+2. **Define some metaprogramming** that specifies how the copyWith() method is
+   defined in terms of the class’s fields. That metaprogramming needs to be
+   authored in some kind of metaprogramming language.
+3. It uses some sort of **introspection mechanism** to walk over the class’s
+   field declarations.
+4. Based on that it uses some other **construction mechanism** to produce a new
+   method including its signature and body. The resulting is then added to the
+   original class declaration.
+
+The approach we propose is **macros**, and the above requirements list can be
+stated more concretely as:
+
+1. We design some sort of syntax to apply a macro by name to a given class,
```

Comment:
This list is too apologetic.  It's ok to be vague, I'd suggest being less concrete.  Just state the problems to be solved.  For example something along the following lines:

1) We propose to design a syntax to apply a macro to a given class, method declaration, etc.
2) We propose to design a way of defining macros which can be then be applied as above.
3) We propose to design an introspection system to allow macros to introspect on the program during their execution.  This might be an imperative API.... etc
4) We propose to consider designing a quotation syntax so that macros that generate Dart code can generate code in a readable fashion and take advantage of existing Dart tooling... etc.